tests/Unit/DataStructures/Tensor/Expressions/Test_Product.cpp

```diff
@@ -0,0 +1,1099 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Framework/TestingFramework.hpp"
+
+#include <cstddef>
+#include <iterator>
+#include <numeric>
+
+#include "DataStructures/Tensor/Expressions/Evaluate.hpp"
+#include "DataStructures/Tensor/Expressions/Product.hpp"
+#include "DataStructures/Tensor/Expressions/TensorExpression.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "Utilities/MakeWithValue.hpp"
+
+namespace {
+template <typename... Ts>
+void create_tensor(gsl::not_null<Tensor<double, Ts...>*> tensor) noexcept {
+  std::iota(tensor->begin(), tensor->end(), 0.0);
+}
+
+template <typename... Ts>
+void create_tensor(gsl::not_null<Tensor<DataVector, Ts...>*> tensor) noexcept {
+  double value = 0.0;
+  for (auto index_it = tensor->begin(); index_it != tensor->end(); index_it++) {
+    for (auto vector_it = index_it->begin(); vector_it != index_it->end();
+         vector_it++) {
+      *vector_it = value;
+      value += 1.0;
+    }
+  }
+}
+
+/// \ingroup TestingFrameworkGroup
+/// \brief Test the outer product of a rank 0 tensor with another tensor is
+/// correctly evaluated
+///
+/// \details
+/// The outer product cases tested are:
+/// - \f$L = R * R\f$
+/// - \f$L = R * R * R\f$
+/// - \f$L^{a} = R * S^{a}\f$
+/// - \f$L_{ai} = R * T_{ai}\f$
+///
+/// For the last two cases, both operand orderings are tested. For the last
+/// case, both LHS index orderings are tested.
+///
+/// \tparam DataType the type of data being stored in the product operands
+template <typename DataType>
+void test_outer_product_rank_0_operand(const DataType& used_for_size) noexcept {
+  Tensor<DataType> R{{{used_for_size}}};
+  if constexpr (std::is_same_v<DataType, double>) {
+    // Instead of the tensor's value being the whole number, `used_for_size`
```

Comment:
I think this is slightly clearer phrased as "Replace tensor's value from `used_for_size` to a proper double" or similar... for me it took reading this twice to understand that by "instead of X" you meant "it is currently X, but I want it to be Y"

(And perhaps below "Replace tensor values so each `DataVector` element has a unique value" ... but if you rename `create_tensor` you might not need a comment there at all. Up to you!)